src/goto-instrument/expand_pointer_predicates.cpp

```diff
@@ -0,0 +1,211 @@
+#include <util/pointer_predicates.h>
+#include <util/simplify_expr.h>
+
+#include <goto-programs/goto_model.h>
+
+class expand_pointer_predicatest
+{
+public:
+  expand_pointer_predicatest(
+    symbol_tablet &_symbol_table,
+    goto_functionst &_goto_functions)
+    : ns(_symbol_table),
+      symbol_table(_symbol_table),
+      goto_functions(_goto_functions),
+      local_bitvector_analysis(nullptr)
+  {
+  }
+
+  void operator()();
+
+protected:
+  namespacet ns;
+  symbol_tablet &symbol_table;
+  goto_functionst &goto_functions;
+  local_bitvector_analysist *local_bitvector_analysis;
+
+  void expand_pointer_predicates();
+
+  void expand_assumption(
+    goto_functionst::goto_functiont &,
+    goto_programt::targett target,
+    exprt &assume_expr);
+
+  void expand_assertion(exprt &assert_expr);
+
+  const symbolt &
+  new_tmp_symbol(const typet &type, const source_locationt &source_location);
+};
+
+bool is_lvalue(const exprt &expr);
+
+void expand_pointer_predicatest::expand_pointer_predicates()
+{
+  Forall_goto_functions(f_it, goto_functions)
+  {
+    goto_functionst::goto_functiont &goto_function = f_it->second;
+    Forall_goto_program_instructions(i_it, goto_function.body)
+    {
+      if(i_it->is_assert())
+      {
+        expand_assertion(i_it->guard);
+      }
+      else if(i_it->is_assume())
+      {
+        expand_assumption(goto_function, i_it, i_it->guard);
+      }
+      else
+      {
+        continue;
+      }
+    }
+  }
+}
+
+void expand_pointer_predicatest::expand_assumption(
+  goto_functionst::goto_functiont &goto_function,
+  goto_programt::targett target,
+  exprt &assume_expr)
+{
+  goto_programt &program = goto_function.body;
+  local_bitvector_analysist local_bitvector_analysis(goto_function, ns);
+  goto_programt assume_code;
+  for(depth_iteratort it = assume_expr.depth_begin();
+      it != assume_expr.depth_end();)
+  {
+    if(it->id() == ID_points_to_valid_memory)
+    {
+      exprt &valid_memory_expr = it.mutate();
+      exprt &pointer_expr = valid_memory_expr.op0();
+      exprt size_expr = valid_memory_expr.op1();
+      simplify(size_expr, ns);
+
+      // This should be forced by typechecking.
+      INVARIANT(
+        pointer_expr.type().id() == ID_pointer && is_lvalue(pointer_expr),
+        "Invalid argument to valid_pointer.");
+
+      local_bitvector_analysist::flagst flags =
+        local_bitvector_analysis.get(target, pointer_expr);
+      // Only create a new object if the pointer may be uninitialized.
+      if(flags.is_uninitialized() || flags.is_unknown())
+      {
+        typet object_type = type_from_size(size_expr, ns);
+
+        // Decl a new variable (which is therefore unconstrained)
+        goto_programt::targett d = assume_code.add_instruction(DECL);
+        d->source_location = assume_expr.source_location();
+        const symbol_exprt obj =
+          new_tmp_symbol(object_type, d->source_location).symbol_expr();
+        d->code = code_declt(obj);
```

Comment:
Please use the recently added API doing `assume_code.add(goto_programt::make_decl(...));` to combine most of the above into a single instruction.